packages/dataviews/src/bulk-actions.js

```diff
@@ -0,0 +1,207 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	privateApis as componentsPrivateApis,
+	Button,
+	Modal,
+} from '@wordpress/components';
+import { __, sprintf, _n } from '@wordpress/i18n';
+import { useMemo, useState, useRef, useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from './lock-unlock';
+
+const {
+	DropdownMenuV2: DropdownMenu,
+	DropdownMenuGroupV2: DropdownMenuGroup,
+	DropdownMenuItemV2: DropdownMenuItem,
+	DropdownMenuSeparatorV2: DropdownMenuSeparator,
+} = unlock( componentsPrivateApis );
+
+function ActionWithModal( {
+	action,
+	selectedItems,
+	setActionWithModal,
+	bulkActionsButtonRef,
+} ) {
+	const eligibleItems = useMemo( () => {
+		return selectedItems.filter( ( item ) => action.isEligible( item ) );
+	}, [ action, selectedItems ] );
+	const { RenderModal, hideModalHeader } = action;
+	const onCloseModal = useCallback( () => {
+		setActionWithModal( undefined );
+		// Focus the bulk actions button.
+		if ( bulkActionsButtonRef?.current?.focus ) {
+			bulkActionsButtonRef.current.focus();
+		}
+	}, [ setActionWithModal, bulkActionsButtonRef ] );
+	return (
+		<Modal
+			title={ ! hideModalHeader && action.label }
+			__experimentalHideHeader={ !! hideModalHeader }
+			onRequestClose={ onCloseModal }
+			overlayClassName="dataviews-action-modal"
+		>
+			<RenderModal items={ eligibleItems } closeModal={ onCloseModal } />
+		</Modal>
+	);
+}
+
+function BulkActionItem( {
+	action,
+	selectedItems,
+	onMenuOpenChange,
+	setActionWithModal,
+	bulkActionsButtonRef,
+} ) {
+	const eligibleItems = useMemo( () => {
+		return selectedItems.filter( ( item ) => action.isEligible( item ) );
+	}, [ action, selectedItems ] );
+	return (
+		<DropdownMenuItem
+			key={ action.id }
+			disabled={ eligibleItems.length === 0 }
+			onClick={ async ( event ) => {
+				event.preventDefault();
```

Comment:
DropdownMenuItem components by default cause the menu to close when clicked. If we want to change this behavior, we should use the `hideOnClick` prop instead of using `event.preventDefault()`.

Reply:
The solution proposed in [this comment](https://github.com/WordPress/gutenberg/pull/57255#discussion_r1442663814) uses `hideOnClick` instead of `event.preventDefault()`

Reply:
Good one, I will look into other places and remove event.preventDefault(). I followed what we were doing in other places and I think hideOnClick may be a better fit there too.